gateway/src/main/java/io/camunda/zeebe/gateway/admin/backup/BackupRequestHandler.java

```diff
@@ -25,40 +31,144 @@ public BackupRequestHandler(final BrokerClient brokerClient) {
 
   @Override
   public CompletableFuture<Long> takeBackup(final long backupId) {
-    final BrokerClusterState topology = topologyManager.getTopology();
-    if (topology == null) {
-      return CompletableFuture.failedFuture(
-          backupFailed(backupId, new NoTopologyAvailableException()));
-    }
-
-    final int expectedPartitionCount = topology.getPartitionsCount();
-    final int knownPartitions = topology.getPartitions().size();
-    if (expectedPartitionCount != knownPartitions) {
+    final Either<Throwable, Boolean> topologyComplete = checkTopologyComplete();
+    if (topologyComplete.isLeft()) {
       return CompletableFuture.failedFuture(
-          backupFailed(
-              backupId,
-              new IncompleteTopologyException(
-                  "Expected to send request to all %d partitions, but found only %d partitions in topology."
-                      .formatted(expectedPartitionCount, knownPartitions))));
+          operationFailed("take", backupId, topologyComplete.getLeft()));
     }
 
     final var backupTriggered =
-        topology.getPartitions().stream()
-            .map(partitionId -> getRequestForPartition(backupId, partitionId))
+        topologyManager.getTopology().getPartitions().stream()
+            .map(partitionId -> getBackupRequestForPartition(backupId, partitionId))
             .map(brokerClient::sendRequestWithRetry)
             .toArray(CompletableFuture[]::new);
 
     return CompletableFuture.allOf(backupTriggered)
         .thenApply(ignore -> backupId)
         .exceptionallyCompose(
-            error -> CompletableFuture.failedFuture(backupFailed(backupId, error.getCause())));
+            error ->
+                CompletableFuture.failedFuture(
+                    operationFailed("take", backupId, error.getCause())));
+  }
+
+  @Override
+  public CompletableFuture<BackupStatus> getStatus(final long backupId) {
+    final Either<Throwable, Boolean> topologyComplete = checkTopologyComplete();
+    if (topologyComplete.isLeft()) {
+      return CompletableFuture.failedFuture(
+          operationFailed("query", backupId, topologyComplete.getLeft()));
+    }
+
+    final var statusesReceived =
+        topologyManager.getTopology().getPartitions().stream()
+            .map(partitionId -> getStatusQueryForPartition(backupId, partitionId))
+            .map(brokerClient::sendRequestWithRetry)
+            .toList();
+
+    return CompletableFuture.allOf(statusesReceived.toArray(CompletableFuture[]::new))
+        .thenApply(ignore -> aggregatePartitionStatus(backupId, statusesReceived))
+        .exceptionallyCompose(
+            error ->
+                CompletableFuture.failedFuture(
+                    operationFailed("query", backupId, error.getCause())));
+  }
+
+  private Either<Throwable, Boolean> checkTopologyComplete() {
```

Comment:
:wrench: I would recommend returning a `CompletionStage<BrokerClusterState>` here instead of `Either`. We don't really care about the return value at the moment, but what we want is "if the topology is complete, use it immediately after; if it's not, fail". As our public API relies on futures, why not simply already return a failed or completed future, and go from there? Even if the operation is not explicitly asynchronous here, it fits better with our asynchronous API, and who knows if one day it might not become asynchronous.

e.g.

```java
  @Override
  public CompletionStage<Long> takeBackup(final long backupId) {
    return checkTopologyComplete()
        .thenApply(topology -> broadcastBackupRequest(backupId, topology))
        .thenApply(ignore -> backupId)
        .exceptionallyCompose(
            error ->
                CompletableFuture.failedFuture(
                    operationFailed("take", backupId, error.getCause())));
  }

  private CompletionStage<BrokerClusterState> checkTopologyComplete() {
    final BrokerClusterState topology = topologyManager.getTopology();
    if (topology == null) {
      return CompletableFuture.failedFuture(new NoTopologyAvailableException());
    }

    final int expectedPartitionCount = topology.getPartitionsCount();
    final int knownPartitions = topology.getPartitions().size();
    if (expectedPartitionCount != knownPartitions) {
      return CompletableFuture.failedFuture(
          new IncompleteTopologyException(
              "Expected to send request to all %d partitions, but found only %d partitions in topology."
                  .formatted(expectedPartitionCount, knownPartitions)));
    }

    return CompletableFuture.completedFuture(topology);
  }

  private CompletableFuture<?>[] broadcastBackupRequest(final long backupId,
      final BrokerClusterState topology) {
    return topology.getPartitions().stream()
        .map(partitionId -> getBackupRequestForPartition(backupId, partitionId))
        .map(brokerClient::sendRequestWithRetry)
        .toArray(CompletableFuture[]::new);
  }
```